File: intel-hex-to-bin.java/src/main/java/demo/Record.java
package demo;

import static com.google.common.base.Preconditions.checkArgument;

/**
 * Intel HEX parsed record.
 */
final public class Record {

    final private short byteCount;
    final private int address;
    final private RecordType recordType;
    final private short[] data;
    final private short checksum;

    public Record(Row row) {
        this.byteCount = row.getUnsigned8Bit();
        this.address = row.getUnsigned16Bit();
        this.recordType = RecordType.of(row.getUnsigned8Bit());

        this.data = new short[this.byteCount];
        for (int i = 0; i < this.byteCount; i++) {
            this.data[i] = row.getUnsigned8Bit();
        }

        this.checksum = row.getUnsigned8Bit();

        checkArgument(row.isEmpty(), "Wrong bytes count in the record: " + this);
        validateChecksum();
    }

    private void validateChecksum() {
        short sum = byteCount;

        sum += address >> 8;
        sum += address & 0xFF;

        sum += recordType.getCode();

        for (int i = 0; i < this.byteCount; i++) {
            sum += this.data[i];
        }

        sum = (short) (((~sum) + 1) & 0xFF);

        checkArgument(sum == this.checksum, String.format("Wrong checksum %s in the record: %s", Hex.to8BitHexString(sum), this.toString()));
    }

    private short getUnsigned8BitFromData(int index) {
        return this.getData()[index];
    }

    public int getUnsigned16BitFromData(int index) {
        checkArgument(this.getData().length == 2);
        return (getUnsigned8BitFromData(index) << 8) + getUnsigned8BitFromData(index + 1);
    }

    public short getByteCount() {
        return byteCount;
    }

    public int getAddress() {
        return address;
    }

    public RecordType getRecordType() {
        return recordType;
    }

    public short[] getData() {
        return data;
    }

    public short getChecksum() {
        return checksum;
    }

    @Override
    public String toString() {
        return "Record{" +
                "byteCount=" + Hex.to8BitHexString(byteCount) +
                ", address=" + Hex.to16BitHexString(address) +
                ", recordType=" + recordType +
                ", data=" + Hex.to8BitHexString(data) +
                ", checksum=" + Hex.to8BitHexString(checksum) +
                '}';
    }

}

File: intel-hex-to-bin.scala/src/main/scala/demo/Record.scala
package demo

import com.google.common.base.Preconditions.checkArgument

/**
  * Intel HEX parsed record.
  */
class Record(private val row: Row) {

  val byteCount: Short = row.getUnsigned8Bit
  val address: Int = row.getUnsigned16Bit
  val recordType: RecordType = RecordTypeConverter.toRecordType(row.getUnsigned8Bit)

  val data: Array[Short] = new Array[Short](this.byteCount)
  for (i <- 0 until this.byteCount) {
    this.data(i) = row.getUnsigned8Bit
  }

  val checksum: Short = row.getUnsigned8Bit

  checkArgument(row.isEmpty, ("Wrong bytes count in the record: " + this).asInstanceOf[Any])
  validateChecksum()

  private def validateChecksum() = {
    var sum: Short = byteCount

    sum = (sum + (address >> 8)).toShort
    sum = (sum + (address & 0xFF)).toShort

    sum = (sum + RecordTypeConverter.toCode(recordType)).toShort

    for (i <- 0 until this.byteCount) {
      sum = (sum + this.data(i)).toShort
    }

    sum = (((sum ^ 0xFF) + 1) & 0xFF).toShort

    checkArgument(sum == this.checksum, java.lang.String.format("Wrong checksum %s for the record: %s", Hex.to16BitHexString(sum), this.toString()).asInstanceOf[Any])
  }

  private def getUnsigned8BitFromData(index: Int): Short = {
    this.data(index)
  }

  def getUnsigned16BitFromData(index: Int): Int = {
    checkArgument(this.data.length == 2)
    (getUnsigned8BitFromData(index) << 8) + getUnsigned8BitFromData(index + 1)
  }

  override def toString: String = {
    "Record{" +
      "byteCount=" + Hex.to8BitHexString(byteCount) +
      ", address=" + Hex.to16BitHexString(address) +
      ", recordType=" + recordType +
      ", data=" + Hex.to8BitHexString(data) +
      ", checksum=" + Hex.to8BitHexString(checksum) +
      '}'
  }

}


